### PR TITLE
Add feature dotted border for Rectangle

### PR DESCRIPTION
**Pull request by inesvar**
Thank you for helping out with embedded-graphics development! Please:

- [X] Check that you've added passing tests and documentation
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) if your changes affect the **public API**
- [X] Run `rustfmt` on the project
- [x] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

Support for dotted/dashed borders and fills was requested in #771. This pull request is only the beginning of the implementation, starting with dotted borders for the `Rectangle` primitive.

### Developer tests

I've tested the existing commits by comparing screenshots of the regular and dotted border (with all stroke alignment styles), see [screenshots](https://github.com/inesvar/test-embedded-graphics/blob/ec6cfb16cfd1ab0ef81f9b06369c83fe1ebb0baf/README.md) (the testing code is in the linked repo).

### Testing issue

I'm not sure what kind of tests should be added for this feature, any advice?
Checking that the dotted border fits the regular border seemed relevant, are there existing tests doing this kind of stuff using screenshots ?

### Scope of the PR

I plan on continuing the implementation of #771 with dotted borders for `Line` and `Polyline`. If you think it would be best to group this in one PR, just let me know.

### CODE REVIEW

**Comment by rfuest** [comment]
Thanks for the PR. I didn't have time to look at the code yet, but I can at least give you some initial feedback.

---

> ### Developer tests
> 
> I've tested the existing commits by comparing screenshots of the regular and dotted border (with all stroke alignment styles), see [screenshots](https://github.com/inesvar/test-embedded-graphics/blob/771-dotted-border-for-rectangle/README.md) (the testing code is in the linked repo).

Have you seen the [`debug-tools`](https://github.com/embedded-graphics/debug-tools) repo? It provides an easy way to write interactive debug tools, which can help to find edge cases that still need work. But the repo might not have been updated to the latest version of e-g yet.

> ### Testing issue
> 
> I'm not sure what kind of tests should be added for this feature, any advice? Checking that the dotted border fits the regular border seemed relevant, are there existing tests doing this kind of stuff using screenshots ?

First I would suggest to add a few tests using [`MockDisplay`](https://docs.rs/embedded-graphics/latest/embedded_graphics/mock_display/index.html), that compare against an expected pattern inside the source code. This should work well for small(ish) border sizes.

Checking against a regular border is a good idea, if you think that 64x64 pixel is enough space you could also use `MockDisplay` for this, otherwise you would need to use `Framebuffer`. For `MockDisplay` the test could look like this:
```rust
let mut regular = MockDisplay::new();
draw_regular_border(&mut regular).unwrap();

let mut dotted = MockDisplay::new();
draw_dotted_border(&mut dotted).unwrap();

for p in regular.bounding_box().points() {
   // checks that the same pixels were drawn in both cases,
   // but ignores the color
   assert_eq!(regular.get_pixel(p).is_some(), dotted.get_pixel(p).is_some());
}
```
`Framebuffer` doesn't track which pixels have been drawn and you would need to pick distinct colors to distinguish between border, dots, fill, and background, but otherwise the test should work pretty much the same.

> ### Scope of the PR
> 
> I plan on continuing the implementation of #771 with dotted borders for `Line` and `Polyline`. If you think it would be best to group this in one PR, just let me know.

Let's keep the PRs focused on one primitive at a time to make reviewing easier.

---

I would like to change the way the dots are a aligned on the border, at least for small border width:
image

The gaps in the center are, in my opinion, more distracting than having some of the corner dots missing:
image

**Comment by inesvar** [comment]
Thanks for your detailed answer.
I will get back to you in the coming week.

Any commits named "temp" are WIP and not meant to be reviewed.

**Comment by inesvar** [comment]
I changed the implementation so that missing corner dots are allowed for border widths of 1px and 2px.
Here are screenshots of various rectangles with the [previous code](https://raw.githubusercontent.com/inesvar/test-embedded-graphics/f386b050925a9e19bb92c875c4abdec744dc7caa/screenshots/various_rectangles_v1.png) and the [new one](https://raw.githubusercontent.com/inesvar/test-embedded-graphics/f386b050925a9e19bb92c875c4abdec744dc7caa/screenshots/various_rectangles_v2.png) .

Let me know what you think, I'll start working on tests in the next week.

**Comment by rfuest** [comment]
I've played with the code a bit and it works well in most cases. But these drawing routines unfortunately have lots of details that need to be considered. Here are some nitpicky comments: 

### Large border sizes with small rectangle sizes

If the length of a vertical or horizontal border is small enough that 3 dots don't fit anymore there is a very large gap between the border dots (right image). If only one pixel more is added the display looks much better (left image). I think we should allow the gaps to shrink a bit to make the jump between 2 and 3 dots less apparent. Having the allowed values for the gap size centered around the ideal value, with some wiggle room in the positive and negative direction, might also improve the spacing in general and not only this case.

image

### Shape of 3 pixel wide borders

Here is a comparison of border width from 1 px to 5 px:

image

In my opinion the 3 pixel wide border doesn't look good an I would never use is. The dots seem to be visually smaller than the chunky 2 pixel ones.

What do you think about using rectangles for border width <= 3 px? Like this:

image

Using rectangles instead of circles for these small sizes could also improve the drawing performance a bit.

### Stroke alignment

At the moment the stroke alignment option doesn't influence how the dots are drawn in relation to the fill. Here is how it should work (from left to right: outside, center, inside):

image

File: src/primitives/primitive_style.rs [hunk 1]
```diff
@@ -42,6 +42,13 @@ where
     /// This property only applies to closed shapes (rectangle, circle, ...) and is
     /// ignored for open shapes (line, ...).
     pub stroke_alignment: StrokeAlignment,
+
+    /// Stroke style.
+    ///
+    /// The `stroke_style` sets a particular style, for instance the dotted style.
+    /// This feature is only implemented for the [`Rectangle`](crate::primitives::Rectangle)
+    /// primitive, and `stroke_style` will be ignored for all other primitives.
+    pub stroke_style: Option<StrokeStyle>,
```
Comment (rfuest) [inline src/primitives/primitive_style.rs:51]: There is no need to use an `Option` here. Just add a `Solid` stroke style and make that the default.

**Comment by inesvar** [comment]
Hello !

I have applied the suggested changes, I do agree that the result is more aesthetic.
That is, except for the stroke alignment, because the code already worked as described.

I didn't have time to update the PR yet, but I did the development and I drew the results in this commit :
https://github.com/inesvar/test-embedded-graphics/commit/5333f34492b06b03b02ccb0050054f720772d615.

screenshots_v5 show the final results, and screenshots_v6 show that the dotted border is correctly aligned with the fill.

You can play with it yourself as the Cargo.toml has a dependency on my git dev branch.

**Comment by rfuest** [comment]
> (That is, except for the stroke alignment, because the code already worked as described).
> ...
> screenshots_v5 show the final results, and screenshots_v6 show that the dotted border is correctly aligned with the fill.

I didn't have time to look at the changes in your latest revision yet, but I think I have an idea why our stroke alignment results were different. Your test draws the fill and stroke separately using two different `PrimitiveStyle`s. But in my test I was using a single `PrimitiveStyle` with both stroke and fill enabled.

**Comment by inesvar** [comment]
Oh I see how it works now. I will update `fill_area` accordingly.

On another subject, I realized that it was a bad idea to use central symmetry in `draw_dotted_rectangle_border_with_dotted_corners` as it will clash with the dotted fill. I will revert to axial symmetry.

I'll be waiting for your appraisal of the code on the dev branch https://github.com/inesvar/embedded-graphics/tree/develop before updating the PR, as I expect you will do further testing before the review. The dev branch should be updated this week-end.

File: src/primitives/primitive_style.rs [hunk 2]
```diff
@@ -118,7 +126,11 @@ where
     /// Returns the fill area.
     pub(in crate::primitives) fn fill_area<P: OffsetOutline>(&self, primitive: &P) -> P {
         // saturate offset at i32::min_value() if stroke width is to large
-        let offset = -self.inside_stroke_width().saturating_as::<i32>();
+        let offset = if self.stroke_style == StrokeStyle::Solid {
+            -self.inside_stroke_width().saturating_as::<i32>()
+        } else {
+            0
```
Comment (rfuest) [inline src/primitives/primitive_style.rs:132]:
```suggestion
            // do not shrink the fill area for non solid borders, because the entire fill
            // area is visible through the gaps in the border            
            0
```

File: src/primitives/primitive_style.rs [hunk 3]
```diff
@@ -287,6 +310,23 @@ impl Default for StrokeAlignment {
...
+/// Stroke style.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[cfg_attr(feature = "defmt", derive(::defmt::Format))]
+#[non_exhaustive]
+pub enum StrokeStyle {
+    /// Solid.
+    Solid,
+    /// Dotted.
+    Dotted,
+}
```
Comment (rfuest) [inline src/primitives/primitive_style.rs:322]:
```suggestion
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Default)]
#[cfg_attr(feature = "defmt", derive(::defmt::Format))]
#[non_exhaustive]
pub enum StrokeStyle {
    /// Solid.
    #[default]
    Solid,
    /// Dotted.
    Dotted,
}
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,7 @@ Embedded Graphics is a `no_std` library for adding graphics features to display
 - [#767](https://github.com/embedded-graphics/embedded-graphics/pull/767) Added `load` and `store` methods to `RawData` trait.
 - [#767](https://github.com/embedded-graphics/embedded-graphics/pull/767) Added `MASK` constant to `RawData` trait.
 - [#768](https://github.com/embedded-graphics/embedded-graphics/pull/768) Added 8bit `Rgb332` support.
+- [#772](https://github.com/embedded-graphics/embedded-graphics/pull/772) Added feature dotted border (through the `stroke_style` helper), implemented for `Rectangle`.
```
Comment (rfuest) [inline CHANGELOG.md:28]: Why do you call `stroke_style` a "helper"?
```suggestion
- [#772](https://github.com/embedded-graphics/embedded-graphics/pull/772) Added `PrimitiveStyle::stroke_style` property to draw dotted borders (currently only supported for `Rectangle`).
```

File: core/src/geometry/point.rs [hunk 5]
```diff
@@ -399,6 +399,16 @@ impl Index<usize> for Point {
...
+impl IndexMut<usize> for Point {
+    fn index_mut(&mut self, idx: usize) -> &mut i32 {
+        match idx {
+            0 => &mut self.x,
+            1 => &mut self.y,
+            _ => panic!("index out of bounds: the len is 2 but the index is {}", idx),
+        }
+    }
+}
+
```
Comment (rfuest) [inline core/src/geometry/point.rs:411]: If we don't need this (see other comment) for this PR I would like to remove it. It might be a useful feature, but if it is implemented for `Point` it should also be implemented for `Size` and I would like to keep this PR focused to the dotted border.

**Review by rfuest — CHANGES_REQUESTED** [review]
The stroke alignment works great now! I couldn't find any more major issues and the PR should almost be ready to be merged.

File: src/primitives/rectangle/styled.rs [hunk 6]
```diff
@@ -69,6 +70,95 @@ impl<C: PixelColor> StyledPixels<PrimitiveStyle<C>> for Rectangle {
...
+            translation[idx] = Into::<i32>::into((dot_offset * Real::from(dot_idx)).round());
+            draw_dot_translated_by(top_left_dot, &translation, style, target)?;
+
+            translation[1 - idx] = border_size[1 - idx] as i32;
+            draw_dot_translated_by(top_left_dot, &translation, style, target)?;
+        }
+    }
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (inesvar) [reply src/primitives/rectangle/styled.rs]: Your suggestion is great, it's applied now (I'm pushing soon).

I also made the following changes that you might have an opinion on :
- adding a second helper function for the second dotted border drawing function
- passed `top_left` by reference in both functions (the Circle/Rectangle dot is now created inside the drawing function)
- updated the second function comment to mention it's using Rectangle

File: core/src/geometry/point.rs (hunk 5 above)
Comment (inesvar) [reply core/src/geometry/point.rs]: It's removed.

**Comment by inesvar** [comment]
`just build` and the CI currently fail, namely because this link https://gitlab.freedesktop.org/xorg/font/misc-misc/-/tree/master returns 503. 
Otherwise the doc, tests and fmt are fine.

**Comment by rfuest** [comment]
> `just build` and the CI currently fail, namely because this link https://gitlab.freedesktop.org/xorg/font/misc-misc/-/tree/master returns 503. Otherwise the doc, tests and fmt are fine.

Don't worry about the broken link and the CI test that fails because of a MSRV issue. I'll look into these errors after this PR is merged.

But this one is a genuine problem with a divide by zero: https://github.com/embedded-graphics/embedded-graphics/actions/runs/13887811895/job/38859234599?pr=772#step:7:586 It seems to only happen when the `fixed_point` feature is enabled:
```shell
# works
cargo test -p embedded-graphics

# broken
cargo test -p embedded-graphics --features=fixed_point
```

**Comment by inesvar** [comment]
It's fixed.

On another note, I noticed the `round` function in geometry/real.rs has a #[allow(unused)] flag. I'm using this function, should I remove the flag?

**Review by rfuest — APPROVED** [review]
> On another note, I noticed the round function in geometry/real.rs has a #[allow(unused)] flag. I'm using this function, should I remove the flag?

That's a good idea. The rest of the PR looks good now and I'll merge it after you've made that change.

**Comment by inesvar** [comment]
It's done :+1:

Next up, if you're amenable I would like to make a PR to lower the number of clippy warnings (there are 28 currently).

That way I can take a small break before implementing the dashed border for the Rectangle, which seems like the natural next step. Or did you have another idea in mind?
